test: Integration test team page

nimat1 opened the pull request:
Added one CMS file team.test.ts
Passed Screen shot 
![image](https://github.com/user-attachments/assets/5ffda896-96ab-4902-8ae2-79c2a7c0e087)

nora-weisser commented on integration-test/tests/cms/team.test.ts:
```diff
@@ -0,0 +1,12 @@
+    expect(data.page).toBeDefined();
+    expect(data.page.title).toBe('Team');
+    expect(data.contact).toBeDefined();
+    expect(data.contact.title).toBe('Contact us');  
```
This validation will cover API contract validation only partially, since it will check only root properties and will not go deep inside JSON structure. I would highly recommend to use schema verification for checking the whole JSON tree.

nimat1 replied:
Hi, do I need to add team.data.ts file as well, as I was thinking the inclusion of actual response data in the data .ts file contains personal data?

nora-weisser replied:
It is better to verify schema rather than response itself. Two reasons: personal information, team info can change frequently in the future.

dricazenck replied:
+1 to what Nora mentioned!

nora-weisser replied:
Please, also refer to the latest PR merged, it will help to separate some common schemas and reuse them. 
https://github.com/Women-Coding-Community/wcc-backend/pull/164

nimat1 commented:
Added one Schema file for validation "team.schema.ts"
![image](https://github.com/user-attachments/assets/85aabeb5-74bf-4fe5-9edf-4e7db7554b40)

nora-weisser commented on integration-test/utils/datafactory/schemas/team.schema.ts:
```diff
@@ -0,0 +1,93 @@
+        teamSchema: {
+            "type": "object",
+            "properties": {
+                "title": {
```
I can see from Schema of Team Page API, that the properties from the first level of JSON include `id`, `page`, `contact`, `membersByType` properties instead of `title`, etc.
<img width="389" alt="Screenshot 2025-01-01 at 11 43 47" src="https://github.com/user-attachments/assets/bfef05db-63d5-4cb0-98c6-9b4944cb90e4" />

nora-weisser commented on integration-test/utils/datafactory/schemas/team.schema.ts:
```diff
@@ -0,0 +1,93 @@
+                    "type": "string",
+                    "minLength": 1
+                },
+          "images": {
```
Let's separate Image Schema in different file and reuse.

nora-weisser commented:
@nimat1 , please review Schema properties for Schema API response, since I can see that not all of them are mentioned.

nimat1 commented:
Thanks, I will address all mentioned issues.

nimat1 commented:
@nora-weisser thanks foe the review, done the changes. 
1. Added new schema file for Images "images.schema.ts"
2. Added all Schema properties for Team API response.
3. remove the commented line from "team.test.ts"
Please do let me know if there is anything missing or needs correction.

nimat1 commented:
Added another change in the "Team.schema.ts" as API has another item added so now 12 items

nora-weisser commented on integration-test/tests/cms/team.test.ts:
```diff
@@ -0,0 +1,19 @@
+    expect(response.status()).toBe(200);
+    const body = await response.json();
+      // schema validation
+      try {
+        validateSchema(teamSchema, body);
+      } catch (e: unknown) {
+        if (e instanceof Error) {
+          console.error(e.message);
+        } else {
+          console.error('An unknown error occurred');
+        }
+      }
```
Please, make sure you throw the error instead just writing logs in console in case of error, as it is done in footer test. 
example:
````
  // schema validation
  try {
    validateSchema(footerSchema, body);
  } catch (e: unknown) {
    if (e instanceof Error) {
      throw new Error(`Schema validation failed: ${e.message}`);
    } else {
      throw new Error('Schema validation failed with an unknown error');
    }
  }
````

nora-weisser commented on integration-test/utils/datafactory/schemas/team.schema.ts:
```diff
@@ -0,0 +1,135 @@
+    }
+}
+}
+}
```
I don't see many fields in that schema, so I created one, which seems to be working, but needs to be tested if itfails properly in all the cases:
````

import { linkSchema } from './link.schema';
import { imagesSchema } from './images.schema';
import { leadershipMemberSchema } from './leadership.member';


export const teamSchema2 = {
    "type": "object",
    "properties": {
        "id": {
            "type": "string",
            "const": "page:TEAM"
        },
        "page": {
            "type": "object",
            "properties": {
                "id": {
                    "type": "string",
                    "minLength": 1
                },
                "title": {
                    "type": "string",
                    "const": "Team"
                },
                "subtitle": {
                    "type": "string",
                    "minLength": 1
                },
                "description": {
                    "type": "string",
                    "minLength": 1,
                },
                "link": { ...linkSchema.definitions.linkSchema },
                "images": {...imagesSchema.definitions.imagesSchema}
            },
            "additionalProperties": false,
            "required": [
                "title"
            ]
        },
        "contact": {
            "type": "object",
            "properties": {
                "title": {
                    "type": "string",
                    "minLength": 1,
                },
                "description": {
                    "type": "string",
                    "minLength": 1,
                },
                "links": {
                    "type": "array",
                    "items": [
                        {
                            "type": "object",
                            "properties": {
                                "type": {
                                    "type": "string",
                                    "enum": ["youtube", "github", "linkedin", "instagram", "facebook", "twitter", "medium", "slack", "meetup", "email", "unknown"," DEFAULT_LINK "]
                                },
                                "link": {
                                    "minLength": 1
                                }
                            },
                            "required": ["type", "link"],
                            "additionalProperties": false,
                            "if": {
                                "properties": {
                                    "type": { "const": "email" }
                                }
                            },
                            "then": {
                                "properties": {
                                    "link": {
                                        "pattern": "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$"  // Custom email regex
                                    }
                                }
                            },
                            "else": {
                                "properties": {
                                    "link": {
                                        "format": "uri",  // Format as URI
                                        "pattern": "^https?://",  // URL pattern
                                        "minLength": 1
                                    }
                                }
                            }
                        }
                    ]
                }
            },
            "required": [
                "title",
                "links"
            ]
        },
        "membersByType": {
            "type": "object",
            "properties": {
                "directors": {
                    "type": "array",
                    "items": {...leadershipMemberSchema.definitions.leadershipMemberSchema}
                },
                "leads": {
                    "type": "array",
                    "items": {...leadershipMemberSchema.definitions.leadershipMemberSchema}
                },
                "evangelists": {
                    "type": "array",
                    "items": {...leadershipMemberSchema.definitions.leadershipMemberSchema}
                },
            },
            "additionalProperties": false,
            "required": [
                "directors",
                "leads",
                "evangelists"
            ]
        }
    },
    "additionalProperties": false,
    "required": [
        "id",
        "page",
        "contact",
        "membersByType"
    ]
}
````

In this schema I separated `leadershipMemberSchema`:
`````
import { imagesSchema } from "./images.schema";


export const leadershipMemberSchema = {
    $ref: '#/definitions/leadershipMemberSchema',
    definitions: {
        leadershipMemberSchema: {
        "type": "object",
        "properties": {
            "fullName": {
                "type": "string",
                "minLength": 1
            },
            "position": {
                "type": "string",
                "minLength": 1
            },
            "email": {
                "type": "string",
                "pattern": "^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$"
            },
            "slackDisplayName": {
                "type": "string",
                "minLength": 1
            },
            "country": {
                "type": "object",
                "properties": {
                    "countryCode": {
                        "type": "string",
                        "minLength": 1
                    },
                    "countryName": {
                        "type": "string",
                        "minLength": 1
                    }
                },
                "required": [
                    "countryCode",
                    "countryName"
                ]
            },
            "city": {
                "type": "string",
                "minLength": 1
            },
            "companyName": {
                "type": "string",
                "minLength": 1
            },
            "images": {...imagesSchema.definitions.imagesSchema},
            "network": {
                "type": "array",
                "items": [
                    {
                        "type": "object",
                        "properties": {
                            "type": {
                                "type": "string"
                            },
                            "link": {
                                "type": "string"
                            }
                        },
                        "required": [
                            "type",
                            "link"
                        ]
                    }
                ]
            }
        },
        "additionalProperties": false,
        "required": [
            "fullName",
            "position",
            "images",
        ]
    }
    },
  };
````

nora-weisser commented:
Also, please, make sure you run prettier formatter, so code formatted in a correct way.

nora-weisser commented on integration-test/utils/datafactory/schemas/team.schema.ts:
```diff
@@ -0,0 +1,125 @@
+                    ' DEFAULT_LINK ',
+                  ],
+                },
+                link: {
+                  minLength: 1,
+                },
```
Type of link property should be forced to have `string` type, otherwise the test doesn't catch the situation when link is number:
````
    "contact": {
      "title": "Contact us",
      "links": [
        {
          "type": "email",
          "link": 12
        },
        {
          "type": "slack",
          "link": "https://join.slack.com/t/womencodingcommunity/shared_invite/zt-2hpjwpx7l-rgceYBIWp6pCiwc0hVsX8A"
        }
      ]
    },
````

nimat1 replied:
Sure, will do. thanks !

nimat1 replied:
@nora-weisser - In Schema value example, I can see a new field is added 
![image](https://github.com/user-attachments/assets/5a12e60a-a2d1-462c-88c3-60c7496da5fd)
Do we need to add any test for that ?

nora-weisser replied:
good catch! please add this one as well

nimat1 commented:
Added   link: {
                  "type": "string",
                },

nimat1 commented:
@nora-weisser All done, thanks !

womencodingcommunity commented on integration-test/tests/cms/team.test.ts:
```diff
@@ -0,0 +1,18 @@
+import { expect, test } from '@playwright/test';
+import { validateSchema } from '@utils/helpers/schema.validation';
+import { teamSchema } from '@utils/datafactory/schemas/team.schema';
+test('GET /api/cms/v1/team returns correct data', async ({ request }) => {
```
please add empty line between imports and the test name.

nimat1 replied:
Done

womencodingcommunity commented on integration-test/utils/datafactory/schemas/team.schema.ts:
```diff
@@ -0,0 +1,153 @@
+      },
+      required: ['title', 'links'],
+    },
+    backgroundcolor: {
```
I don't think this file is formatted properly. Could you have a look again please?

nimat1 replied:
@womencodingcommunity  done
